crates/core/src/error/summary.rs

```diff
@@ -0,0 +1,617 @@
+//!
+//! # Separation of Concerns
+//!
+//! Critical runtime errors that prevent further execution are handled through the
+//! existing `RusticError` system. The `Summary` will only collect information for
+//! non-fatal events.
+//!
+//! # Compatibility with Existing Error Handling
+//!
+//! Summaries must coexist with error propagation rules. They will not replace
+//! the core behavior of error propagation but act as a complementary mechanism
+//! for presenting non-fatal feedback.
+//!
+//! # User-Friendly Reporting
+//!
+//! Summaries should aggregate detailed runtime information—such as warnings,
+//! issues, and metrics — in a clear and condensed format for the end-user.
+//!
+//! # Aggregation & Condensation
+//!
+//! Similar or repeated errors should be aggregated to avoid redundant information,
+//! presenting users with a high-level overview.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt::{self, Display, Write},
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use ecow::EcoString;
+
+use crate::error::{ErrorKind, RusticError, RusticResult};
+
+pub type Issues = BTreeMap<IssueCategory, BTreeMap<IssueScope, CondensedIssue>>;
+
+pub type Metrics = BTreeMap<EcoString, EcoString>;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Default,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    serde::Serialize,
+)]
+pub enum IssueCategory {
+    #[default]
+    Warning,
+    Error,
+    Info,
+}
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Default,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    serde::Serialize,
+)]
+pub enum IssueScope {
+    #[default]
+    Internal,
+    Unknown,
+    UserInput,
+}
+
+#[derive(Debug, Clone, Default, serde::Serialize)]
+pub struct CondensedIssue {
+    /// High-level description of the problem
+    message: EcoString,
+
+    category: IssueCategory,
+
+    /// Number of occurrences
+    count: usize,
+
+    /// Optional diagnostic information, e.g. an error message
+    root_cause: Option<EcoString>,
+}
+
+#[derive(
+    Debug, Clone, Copy, Default, Hash, PartialEq, Eq, derive_more::Display, serde::Serialize,
+)]
+pub enum DisplayOptionKind {
+    #[default]
+    Issues,
+    Timing,
+    Metrics,
+    All,
+}
+
+#[derive(Debug, Clone)]
+pub struct Summary {
+    /// Name of the active context, e.g. a command or operation
+    context: EcoString,
+
+    /// Start time of the collection
+    // Instant cannot be (de-)serialized, for an implementation see:
+    // https://github.com/serde-rs/serde/issues/1375#issuecomment-419688068
+    start_time: Instant,
+
+    /// End time, when the collection is completed
+    // Serialization: See note above
+    end_time: Option<Instant>,
+
+    /// Collection of non-critical warnings   
+    issues: Issues,
+
+    /// Optional custom metrics collected during execution
+    metrics: Metrics,
+
+    /// Display this data
+    display: HashSet<DisplayOptionKind>,
```

Comment:
As I'm beginning to read through the code, I am not (yet) fully undertanding what  `DisplayOptionKind` is about.
Could you detail it a bit more in the code-doc? (for instance in the enum's doc)